### ISSUE TITLE
Document absURL and relURL’s behaviours with leading `/` slash

### ISSUE THREAD

**Issue by anthonyfok** [gohugoio/hugoDocs#469]
absURL ~and relURL~ seems to ignore the sub-folder of the baseURL when it is given a path with a leading `/` slash, but that behaviour isn't documented in https://gohugo.io/functions/absurl https://gohugo.io/functions/relurl.

See budparr/gohugo-theme-ananke#103 for example.

TODO: Study AbsURL() and RelURL() in helpers/url.go (and maybe fix the comments there too) before touching hugoDocs

**Comment by kaushalmodi** [comment]
This has come up many times before.. I believe it's a bug.. but we need to document whatever the current behavior is.

Here's a [recent thread][1] with lots of detailed discussion.

[1]: https://discourse.gohugo.io/t/baseurl-path-subdir-sometimes-ignored-by-relurl-absurl/11512/5?u=kaushalmodi

**Comment by anthonyfok** [comment]
Okay, here is a historical account of what led to today's absURL behaviour:

1. Hugo 0.14: absURL and relURL were born:

    * In May 2015, @bep added absURL (gohugoio/hugo#1106) and relURL (gohugoio/hugo#1126) and documented them.  At the time, both template functions had consistent behaviours, and gave identical output whether a leading `/` slash is present or absent.

2. Hugo 0.15: Same behaviour as in Hugo 0.14.

3. Hugo 0.16: absURL's behaviour changed, and remained the same to this day:

    * In Nov 2015, @mjanda filed gohugoio/hugo#1540 (Make absURL properly handle baseURL with path component) and gohugoio/hugo#1542 (Apply absURL on social tag urls).  The latter was merged first and entered Hugo 0.15, while the former was merged as gohugoio/hugo@0962470850ed6802b645d5bd6663db60807c8f5b a several months later and landed in Hugo 0.16.

    * @mjanda gave the reason for changing absURL not in PR#1540 but in PR#1542:
        > I use baseURL with path component (http://example.com/blog/) and reference my images like /blog/xxx/xxx.png etc.
        >
        > This PR together with #1540 will allow me to use those relative paths in images frontmatter array.

    * From the test cases that @mjanda wrote in gohugoio/hugo@0962470850ed6802b645d5bd6663db60807c8f5b, it can be seen that:
        * Given baseURL = "http://base/path":
            * `{{ absURL "test/2/foo/" }}` gives `http://base/path/test/2/foo/`
            * `{{ absURL "/test/2/foo/" }}` gives `http://base/test/2/foo/`

    * The documentation was not updated to reflect this change, and gohugoio/hugo#1540 didn't make it to the Hugo 0.16 release notes either.

    * No corresponding change was made to relURL, hence `{{ absURL "/test/2/foo" }}` and `{{ relURL "/test/2/foo" }}` produce different results, for better or for worse.

**Comment by anthonyfok** [comment]
Extensive discussion `HUGO_CANONIFYURLS=true` here: https://discourse.gohugo.io/t/baseurl-path-subdir-sometimes-ignored-by-relurl-absurl/11512 (Thank you @kaushalmodi for providing that link!)

**Comment by kaushalmodi** [comment]
@anthonyfok You're welcome. Here's a test page (that I also link in the discourse thread you linked), that summarizes the absURL vs relURL vs canonifyURLs: https://hugo-sandbox.netlify.com/subdir-canonify-true/.

This inconsistency definitely looks like a bug to me. I hope it gets fixed.

**Comment by bep** [comment]
Not sure what the issue is, but I just wanted to point out that:

* When canonifyURLs = true, we make absolute URLs out of relative ones from **the rendered HTML output**
** Since the above is a stupid search and replace that prepends the `baseURL`, the output from `relURL` cannot have any sub-paths in it if this is going to work.

So, whatver you do to verify "correctness", if will have to be the rendered output (and not some debug output of `relURL`. 

Hope that helps.

**Comment by anthonyfok** [comment]
Thank you for the clarification, @ bep!

> Not sure what the issue is...

There are actually a few separate issues that got kind of mixed into one single GitHub issue about clarifying absURL/relURL documentation:

1. @mjanda's change to absURL (leading `/` slash) in Hugo 0.16 needs to be documented.

2. (canonifyURLs = false) relURL did not get the same change about leading `/` slash (which I just realized today), and we should probably discuss in a new issue over at gohugoio/hugo whether relURL should also drop the baseURL sub-path component like absURL does when given a leading `/` slash.

3. relURL's behaviour when `canonifyURLs = true`, which @kaushalmodi and you clarified.  It is again an entirely different issue, which I just realized about an hour ago.

So, yes, it suddenly became a lot more complicated than I thought.  (I went dumbfounded... my head sort of exploded.)  But your clairification did ease my fear a lot.  :-)

Moving forward:

* We should document the current absURL and relURL behaviours, especially those which catches users by surprise.
* In the more distant future, should we consider changing relURL's behaviours (regarding points 2 and 3)?  [To be discussed]
